File: util/src/com/aicp/gear/util/DeviceUtils.java
/*
* Copyright (C) 2014 SlimRoms Project
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
* http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
package com.aicp.gear.util;

import static android.view.DisplayCutout.BOUNDS_POSITION_LEFT;
import static android.view.DisplayCutout.BOUNDS_POSITION_RIGHT;

import android.bluetooth.BluetoothAdapter;
import android.content.ContentResolver;
import android.content.Context;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.content.res.Resources;
import android.graphics.Point;
import android.graphics.Rect;
import android.hardware.camera2.CameraAccessException;
import android.hardware.camera2.CameraCharacteristics;
import android.hardware.camera2.CameraManager;
import android.hardware.display.DisplayManager;
import android.hardware.display.WifiDisplayStatus;
import android.net.ConnectivityManager;
import android.nfc.NfcAdapter;
import android.provider.Settings;
import android.os.Vibrator;
import android.telephony.TelephonyManager;
import android.util.DisplayMetrics;
import android.view.DisplayCutout;
import android.view.DisplayInfo;
import android.view.WindowManager;
import android.util.Log;

import com.android.internal.telephony.PhoneConstants;

import java.util.ArrayList;
import java.util.List;

public class DeviceUtils {

    private static final String TAG = "AicpGearDeviceUtils";

    private static final boolean DEBUG = false;

    private static final String SETTINGS_METADATA_NAME = "com.android.settings";

    // Device types
    private static final int DEVICE_PHONE  = 0;
    private static final int DEVICE_HYBRID = 1;
    private static final int DEVICE_TABLET = 2;

    private static final int NO_CUTOUT = -1;

    public static boolean deviceSupportsRemoteDisplay(Context ctx) {
        DisplayManager dm = (DisplayManager) ctx.getSystemService(Context.DISPLAY_SERVICE);
        return (dm.getWifiDisplayStatus().getFeatureState()
                != WifiDisplayStatus.FEATURE_STATE_UNAVAILABLE);
    }
/*
    public static boolean deviceSupportsUsbTether(Context context) {
        ConnectivityManager cm =
            (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
        return (cm.getTetherableUsbRegexs().length != 0);
    }

    public static boolean deviceSupportsMobileData(Context context) {
        ConnectivityManager cm =
            (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
        return cm.isNetworkSupported(ConnectivityManager.TYPE_MOBILE);
    }
*/
    public static boolean deviceSupportsBluetooth() {
        return (BluetoothAdapter.getDefaultAdapter() != null);
    }

    public static boolean deviceSupportsNfc(Context context) {
        return NfcAdapter.getDefaultAdapter(context) != null;
    }
/*
    public static boolean deviceSupportsLte(Context context) {
        final TelephonyManager tm =
            (TelephonyManager) context.getSystemService(Context.TELEPHONY_SERVICE);
        return (tm.getLteOnCdmaMode() == PhoneConstants.LTE_ON_CDMA_TRUE);
                    // || tm.getLteOnGsmMode() != 0; // add back if when we have support on LP for it
    }
*/
    public static boolean deviceSupportsGps(Context context) {
        return context.getPackageManager().hasSystemFeature(PackageManager.FEATURE_LOCATION_GPS);
    }

    public static boolean adbEnabled(ContentResolver resolver) {
            return (Settings.Global.getInt(resolver, Settings.Global.ADB_ENABLED, 0)) == 1;
    }

    public static boolean deviceSupportsVibrator(Context ctx) {
        Vibrator vibrator = (Vibrator) ctx.getSystemService(Context.VIBRATOR_SERVICE);
        return vibrator.hasVibrator();
    }

    public static boolean deviceSupportsTorch(Context context) {
        // Need to be adapted to new torch API
        return true;
    }

    public static FilteredDeviceFeaturesArray filterUnsupportedDeviceFeatures(Context context,
            String[] valuesArray, String[] entriesArray) {
        if (valuesArray == null || entriesArray == null || context == null) {
            return null;
        }
        List<String> finalEntries = new ArrayList<String>();
        List<String> finalValues = new ArrayList<String>();
        FilteredDeviceFeaturesArray filteredDeviceFeaturesArray =
            new FilteredDeviceFeaturesArray();

        for (int i = 0; i < valuesArray.length; i++) {
            if (isSupportedFeature(context, valuesArray[i])) {
                finalEntries.add(entriesArray[i]);
                finalValues.add(valuesArray[i]);
            }
        }
        filteredDeviceFeaturesArray.entries =
            finalEntries.toArray(new String[finalEntries.size()]);
        filteredDeviceFeaturesArray.values =
            finalValues.toArray(new String[finalValues.size()]);
        return filteredDeviceFeaturesArray;
    }

    private static boolean isSupportedFeature(Context context, String action) {
        if (action.equals(ActionConstants.ACTION_TORCH)
                        && !deviceSupportsTorch(context)
                || action.equals(ActionConstants.ACTION_VIB)
                        && !deviceSupportsVibrator(context)
                || action.equals(ActionConstants.ACTION_VIB_SILENT)
                        && !deviceSupportsVibrator(context)) {
            return false;
        }
        return true;
    }

    public static class FilteredDeviceFeaturesArray {
        public String[] entries;
        public String[] values;
    }

    private static int getScreenType(Context con) {
        WindowManager wm = (WindowManager)con.getSystemService(Context.WINDOW_SERVICE);
        DisplayInfo outDisplayInfo = new DisplayInfo();
        wm.getDefaultDisplay().getDisplayInfo(outDisplayInfo);
        int shortSize = Math.min(outDisplayInfo.logicalHeight, outDisplayInfo.logicalWidth);
        int shortSizeDp =
            shortSize * DisplayMetrics.DENSITY_DEFAULT / outDisplayInfo.logicalDensityDpi;
        if (shortSizeDp < 600) {
            return DEVICE_PHONE;
        } else if (shortSizeDp < 720) {
            return DEVICE_HYBRID;
        } else {
            return DEVICE_TABLET;
        }
    }

    public static boolean isPhone(Context con) {
        return getScreenType(con) == DEVICE_PHONE;
    }

    public static boolean isHybrid(Context con) {
        return getScreenType(con) == DEVICE_HYBRID;
    }

    public static boolean isTablet(Context con) {
        return getScreenType(con) == DEVICE_TABLET;
    }

    /**
     * Returns whether the device is voice-capable (meaning, it is also a phone).
     */
    public static boolean isVoiceCapable(Context context) {
        TelephonyManager telephony =
                (TelephonyManager) context.getSystemService(Context.TELEPHONY_SERVICE);
        return telephony != null && telephony.isVoiceCapable();
    }
/*
    public static boolean isWifiOnly(Context context) {
        ConnectivityManager cm = (ConnectivityManager)context.getSystemService(
                Context.CONNECTIVITY_SERVICE);
        return (cm.isNetworkSupported(ConnectivityManager.TYPE_MOBILE) == false);
    }
*/
    public static boolean deviceSupportsFingerPrint(Context context) {
        PackageManager pm =
          (PackageManager) context.getPackageManager();
        return pm != null && pm.hasSystemFeature(PackageManager.FEATURE_FINGERPRINT);
    }

    public static boolean deviceSupportsFlashLight(Context context) {
        CameraManager cameraManager = (CameraManager) context.getSystemService(
                Context.CAMERA_SERVICE);
        try {
            String[] ids = cameraManager.getCameraIdList();
            for (String id : ids) {
                CameraCharacteristics c = cameraManager.getCameraCharacteristics(id);
                Boolean flashAvailable = c.get(CameraCharacteristics.FLASH_INFO_AVAILABLE);
                Integer lensFacing = c.get(CameraCharacteristics.LENS_FACING);
                if (flashAvailable != null
                        && flashAvailable
                        && lensFacing != null
                        && lensFacing == CameraCharacteristics.LENS_FACING_BACK) {
                    return true;
                }
            }
        } catch (CameraAccessException | AssertionError e) {
            // Ignore
        }
        return false;
    }

    public static int getCutoutType(Context context) {
        final DisplayInfo info = new DisplayInfo();
        context.getDisplay().getDisplayInfo(info);
        final DisplayCutout cutout = info.displayCutout;
        if (cutout == null) {
            if (DEBUG) Log.v(TAG, "noCutout");
            return NO_CUTOUT;
        }
        final Point displaySize = new Point();
        context.getDisplay().getRealSize(displaySize);
        List<Rect> cutOutBounds = cutout.getBoundingRects();
        if (cutOutBounds != null) {
            for (Rect cutOutRect : cutOutBounds) {
                if (DEBUG) Log.v(TAG, "cutout left= " + cutOutRect.left);
                if (DEBUG) Log.v(TAG, "cutout right= " + cutOutRect.right);
                if (cutOutRect.left == 0 && cutOutRect.right > 0) {  //cutout is located on top left
                    if (DEBUG) Log.v(TAG, "cutout position= " + BOUNDS_POSITION_LEFT);
                    return BOUNDS_POSITION_LEFT;
                } else if (cutOutRect.right == displaySize.x && (displaySize.x - cutOutRect.left) > 0) {  //cutout is located on top right
                    if (DEBUG) Log.v(TAG, "cutout position= " + BOUNDS_POSITION_RIGHT);
                    return BOUNDS_POSITION_RIGHT;
                }
            }
        }
        return NO_CUTOUT;
    }
}

File: util/src/com/aicp/gear/util/DeviceUtils.kt
package com.aicp.gear.util

import android.bluetooth.BluetoothAdapter
import android.content.ContentResolver
import android.content.Context
import android.content.pm.PackageManager
import android.hardware.camera2.CameraAccessException
import android.hardware.camera2.CameraCharacteristics
import android.hardware.camera2.CameraManager
import android.hardware.display.DisplayManager
import android.hardware.display.WifiDisplayStatus
import android.nfc.NfcAdapter
import android.os.Vibrator
import android.provider.Settings
import android.telephony.TelephonyManager
import android.util.DisplayMetrics
import android.util.Log
import android.view.DisplayCutout
import android.view.DisplayInfo
import android.view.WindowManager
import java.util.*

object DeviceUtils {
    private const val TAG = "AicpGearDeviceUtils"
    private const val DEBUG = false
    private const val SETTINGS_METADATA_NAME = "com.android.settings"

    // Device types
    private const val DEVICE_PHONE = 0
    private const val DEVICE_HYBRID = 1
    private const val DEVICE_TABLET = 2
    private const val NO_CUTOUT = -1

    @JvmStatic
    fun deviceSupportsRemoteDisplay(ctx: Context): Boolean {
        val dm = ctx.getSystemService(Context.DISPLAY_SERVICE) as DisplayManager
        return dm.wifiDisplayStatus.featureState != WifiDisplayStatus.FEATURE_STATE_UNAVAILABLE
    }

    @JvmStatic
    fun deviceSupportsBluetooth(): Boolean {
        return BluetoothAdapter.getDefaultAdapter() != null
    }

    @JvmStatic
    fun deviceSupportsNfc(context: Context): Boolean {
        return NfcAdapter.getDefaultAdapter(context) != null
    }

    @JvmStatic
    fun deviceSupportsGps(context: Context): Boolean {
        return context.packageManager.hasSystemFeature(PackageManager.FEATURE_LOCATION_GPS)
    }

    @JvmStatic
    fun adbEnabled(resolver: ContentResolver): Boolean {
        return Settings.Global.getInt(resolver, Settings.Global.ADB_ENABLED, 0) == 1
    }

    @JvmStatic
    fun deviceSupportsVibrator(ctx: Context): Boolean {
        val vibrator = ctx.getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
        return vibrator.hasVibrator()
    }

    @JvmStatic
    fun deviceSupportsTorch(context: Context): Boolean {
        // Need to be adapted to new torch API
        return true
    }

    @JvmStatic
    fun filterUnsupportedDeviceFeatures(
        context: Context,
        valuesArray: Array<String>?,
        entriesArray: Array<String>?
    ): FilteredDeviceFeaturesArray? {
        if (valuesArray == null || entriesArray == null) {
            return null
        }
        val finalEntries: MutableList<String> = ArrayList()
        val finalValues: MutableList<String> = ArrayList()
        for (i in valuesArray.indices) {
            if (isSupportedFeature(context, valuesArray[i])) {
                finalEntries.add(entriesArray[i])
                finalValues.add(valuesArray[i])
            }
        }
        return FilteredDeviceFeaturesArray(
            finalEntries.toTypedArray(),
            finalValues.toTypedArray()
        )
    }

    private fun isSupportedFeature(context: Context, action: String): Boolean {
        return !(action == ActionConstants.ACTION_TORCH && !deviceSupportsTorch(context) ||
                action == ActionConstants.ACTION_VIB && !deviceSupportsVibrator(context) ||
                action == ActionConstants.ACTION_VIB_SILENT && !deviceSupportsVibrator(context))
    }

    class FilteredDeviceFeaturesArray(
        var entries: Array<String>,
        var values: Array<String>
    )

    private fun getScreenType(con: Context): Int {
        val wm = con.getSystemService(Context.WINDOW_SERVICE) as WindowManager
        val outDisplayInfo = DisplayInfo()
        wm.defaultDisplay.getDisplayInfo(outDisplayInfo)
        val shortSize = Math.min(outDisplayInfo.logicalHeight, outDisplayInfo.logicalWidth)
        val shortSizeDp = shortSize * DisplayMetrics.DENSITY_DEFAULT / outDisplayInfo.logicalDensityDpi
        return if (shortSizeDp < 600) {
            DEVICE_PHONE
        } else if (shortSizeDp < 720) {
            DEVICE_HYBRID
        } else {
            DEVICE_TABLET
        }
    }

    @JvmStatic
    fun isPhone(con: Context): Boolean {
        return getScreenType(con) == DEVICE_PHONE
    }

    @JvmStatic
    fun isHybrid(con: Context): Boolean {
        return getScreenType(con) == DEVICE_HYBRID
    }

    @JvmStatic
    fun isTablet(con: Context): Boolean {
        return getScreenType(con) == DEVICE_TABLET
    }

    @JvmStatic
    fun isVoiceCapable(context: Context): Boolean {
        val telephony = context.getSystemService(Context.TELEPHONY_SERVICE) as TelephonyManager
        return telephony.isVoiceCapable
    }

    @JvmStatic
    fun deviceSupportsFingerPrint(context: Context): Boolean {
        return context.packageManager.hasSystemFeature(PackageManager.FEATURE_FINGERPRINT)
    }

    @JvmStatic
    fun deviceSupportsFlashLight(context: Context): Boolean {
        val cameraManager = context.getSystemService(Context.CAMERA_SERVICE) as CameraManager
        return try {
            val ids = cameraManager.cameraIdList
            for (id in ids) {
                val c = cameraManager.getCameraCharacteristics(id)
                val flashAvailable = c.get(CameraCharacteristics.FLASH_INFO_AVAILABLE)
                val lensFacing = c.get(CameraCharacteristics.LENS_FACING)
                if (flashAvailable != null && flashAvailable &&
                    lensFacing != null && lensFacing == CameraCharacteristics.LENS_FACING_BACK) {
                    return true
                }
            }
            false
        } catch (e: CameraAccessException) {
            false
        } catch (e: AssertionError) {
            false
        }
    }

    @JvmStatic
    fun getCutoutType(context: Context): Int {
        val info = DisplayInfo()
        context.display?.getDisplayInfo(info)
        val cutout = info.displayCutout
        if (cutout == null) {
            if (DEBUG) Log.v(TAG, "noCutout")
            return NO_CUTOUT
        }
        val displaySize = android.graphics.Point()
        context.display?.getRealSize(displaySize)
        val cutOutBounds = cutout.boundingRects
        if (cutOutBounds != null) {
            for (cutOutRect in cutOutBounds) {
                if (DEBUG) Log.v(TAG, "cutout left= " + cutOutRect.left)
                if (DEBUG) Log.v(TAG, "cutout right= " + cutOutRect.right)
                if (cutOutRect.left == 0 && cutOutRect.right > 0) {
                    if (DEBUG) Log.v(TAG, "cutout position= " + DisplayCutout.BOUNDS_POSITION_LEFT)
                    return DisplayCutout.BOUNDS_POSITION_LEFT
                } else if (cutOutRect.right == displaySize.x && (displaySize.x - cutOutRect.left) > 0) {
                    if (DEBUG) Log.v(TAG, "cutout position= " + DisplayCutout.BOUNDS_POSITION_RIGHT)
                    return DisplayCutout.BOUNDS_POSITION_RIGHT
                }
            }
        }
        return NO_CUTOUT
    }
}

